reasoner/resolution/resolver/DisjunctionResolver.java

```diff
@@ -0,0 +1,220 @@
+        LOG.debug("{}: Updating ResponseProducer for iteration '{}'", name(), newIteration);
+
+        assert newIteration > responseProducerPrevious.iteration();
+        ResponseProducer responseProducerNewIter = responseProducerPrevious.newIterationRetainDedup(Iterators.empty(), newIteration);
+        for (Actor<ConjunctionResolver.Nested> conjunctionResolver : downstreamResolvers) {
+            Filtered downstream = fromUpstream.partialAnswer()
+                    .filterToDownstream(conjunctionRetrievedIds(conjunctionResolver));
+            Request request = Request.create(self(), conjunctionResolver, downstream);
+            responseProducerNewIter.addDownstreamProducer(request);
+        }
+        return responseProducerNewIter;
+    }
+
+    protected Set<Identifier.Variable.Retrievable> conjunctionRetrievedIds(Actor<ConjunctionResolver.Nested> conjunctionResolver) {
+        // TODO use a map from resolvable to resolvers, then we don't have to reach into the state and use the conjunction
+        return iterate(conjunctionResolver.state.conjunction.variables()).filter(v -> v.id().isRetrievable())
+                .map(v -> v.id().asRetrievable()).toSet();
+    }
+
+    @Override
+    protected ResponseProducer responseProducerCreate(Request fromUpstream, int iteration) {
+        LOG.debug("{}: Creating a new ResponseProducer for request: {}", name(), fromUpstream);
+        assert fromUpstream.partialAnswer().isFiltered() || fromUpstream.partialAnswer().isIdentity();
+        ResponseProducer responseProducer = new ResponseProducer(Iterators.empty(), iteration);
+        assert !downstreamResolvers.isEmpty();
+        for (Actor<ConjunctionResolver.Nested> conjunctionResolver : downstreamResolvers) {
+            Filtered downstream = fromUpstream.partialAnswer()
+                    .filterToDownstream(conjunctionRetrievedIds(conjunctionResolver));
+            Request request = Request.create(self(), conjunctionResolver, downstream);
+            responseProducer.addDownstreamProducer(request);
+        }
+        return responseProducer;
+    }
+
+    public static class Nested extends DisjunctionResolver<Nested> {
+
+        public Nested(Actor<Nested> self, grakn.core.pattern.Disjunction disjunction,
+                      Actor<ResolutionRecorder> resolutionRecorder, ResolverRegistry registry,
+                      TraversalEngine traversalEngine, ConceptManager conceptMgr, boolean explanations) {
+            super(self, Nested.class.getSimpleName() + "(pattern: " + disjunction + ")", disjunction,
+                  resolutionRecorder, registry, traversalEngine, conceptMgr, explanations);
+        }
+
+        @Override
+        protected void nextAnswer(Request fromUpstream, ResponseProducer responseProducer, int iteration) {
+            if (responseProducer.hasUpstreamAnswer()) {
+                Partial<?> upstreamAnswer = responseProducer.upstreamAnswers().next();
```

Comment:
im pretty sure this should be possible since we don't execute traversals in disjunctions anymore, we should throw an ILLEGAL_STATE and clean this up